README.md
# 基于antd和umi的前端架构设计

* [技术选型](#技术选型)
* [环境搭建](#环境搭建)
* [启动说明](#启动说明)
* [命令](#命令)
* [发布](#发布)
* [目录结构](#目录结构)
* [路由和菜单](#路由和菜单) 
* [权限管理](#权限管理)     
* [数据mock](#数据mock) 
* [国际化](#国际化) 
* [数据流方案](#数据流方案) 
* [组件](#组件) 
    * [小颗粒组件](#小颗粒组件) 
    * [大颗粒组件](#大颗粒组件) 
* [兼容性](#兼容性) 
* [主题配置](#主题配置) 
* [异步方法封装](#异步方法封装) 
* [动态加载组件](#动态加载组件) 
* [umi约定](#umi约定) 
* [z-index](#z-index) 
* [自定义图标](#自定义图标) 
* [代码规范](#代码规范) 
* [eslint](#eslint) 
* [常见eslint错误](#常见eslint错误) 
* [集成与被集成](#集成与被集成) 
* [样式主题](#样式主题) 
* [其他](#其他)

## 技术选型

| 类型       | 名称                                                                                                                                                                                                                                   |
| :--------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| MVVM       | [react@16.8.3](https://reactjs.org/)                                                                                                                                                                                                   |
| 视图框架   | [antd@3.13.0](https://ant.design/index-cn)、[pro-antd@2.0.0](https://pro.ant.design/index-cn)                                                                                                                                          |
| 数据流方案 | [dva@2.0](https://dvajs.com)                                                                                                                                                                                                           |
| 构建工具   | [umi@2.6.13](http://umijs.org)                                                                                                                                                                                                         |
| 其     他  | [less](http://www.css88.com/doc/less/)、[css modules](http://www.ruanyifeng.com/blog/2016/06/css_modules.html)、[lodash](http://www.css88.com/doc/lodash/)、[jquery](http://www.css88.com/jqapi-1.9/)、[moment](https://momentjs.com/) |



## 环境搭建

1. 安装 [NodeJS](https://nodejs.org/en/)。__确保 node 版本是 >=10.0.0 。

安装完成后，执行下面的命令确认是否安装成功。

```
node -v
npm -v
```

__统一__ 使用 yarn 管理 npm 依赖：

```
npm i yarn -g
```

2. 下载依赖包

```
yarn install
```

注：如果用的是svn,install成功后会在目录下生成`yarn-offline-mirror`和`node_modules` 文件夹，记得把他们添加到忽略名单，无需提交到版本库。git无需关注。

3. 安装IDE插件

以vscode 为例，需安装下列3个插件

+ Prettier - Code formatter 插件
+ EditorConfig for VS Code 插件

至此，环境配置完毕！

## 启动说明

1.  `npm run dev`启动
2. 登录账号密码  admin/123

## 命令

```
npm run dev             // 以开发模式启动
npm run dev:no-mock     // 以开发模式启动(无mock)
npm run prettier        // 格式化src下的.js .jsx .ts .less，以及config和scripts下的.js文件并保存
npm run analyze         // 分析bundle构成,分析依赖模块的体积分布
npm run build           // 打包生产代码
npm run build API_HOST=http://www.baidu.com  // 打包时request 会取这里的值来补全apiurl
npm run dev API_HOST=http://www.baidu.com  // 打包时request 会取这里的值来补全apiurl
npm run lint           // 检验src下的.js .jsx .ts .less
npm run lint:fix       // 检验并修复src下的.js .jsx .ts .less
```

## 发布

命令行执行`npm run build`，会在根目录生成`dist`文件夹。把`dist`提交到服务器即可。

注：在编译前，可以通过`npm run build API_HOST=http://www.baidu.com` 指定异步请求的host部分。具体逻辑如下：

```javascript
{
  url: /^http(s)?:\/\//.test(url) ? url : window['API_HOST'] + url
}
```

## 目录结构

 🔒 表示文件（夹）名是umi的约定关键字

```javascript
|--- config                              🔒
|    |--- config.js                      🔒 umi 配置，同 .umirc.js，二选一   
|--- mock                                🔒 mock 文件所在目录
|--- public                              🔒
|--- dist                                # 源码编译生成的目录                  
|--- src                                 🔒 源码目录                  
|    |--- pages                          🔒
|    |    |--- Log                       # 业务模块1：日志管理
|    |    |    |--- models               🔒 
|    |    |    |    |--- log.js          # 业务级的models是局部作用域，数据不涉及模块间共享时，建议放在模块目录下
|    |    |    |--- services             🔒 
|    |    |    |    |--- log.js          # 业务级 异步请求服务
|    |    |    |--- css
|    |    |    |    |--- index.less      # 业务级样式
|    |    |    |--- img
|    |    |    |    |--- xx.jpg          # 业务级图片资源
|    |    |    |--- index.js             # 视图
|    |    |    |--- _mock.js             🔒 业务级mock
|    |    |    |--- ...
|    |    |--- Tasks                     # 业务模块2
|    |    |--- User                      # 业务模块3
|    |    |--- document.ejs              # HTML模板
|    |--- layouts                        # 布局模板                
|    |    |--- BasicLayout.js                   
|    |    |--- UserLayout.js               
|    |--- components                     # pro-antd组件                
|    |    |--- Charts                   
|    |    |--- Exception               
|    |--- assets                         # 静态资源 
|    |--- utils                          # 工具方法                
|    |--- models                         🔒 数据模型                
|    |--- services                       🔒 异步请求服务                
|    |--- locales                        🔒 国际化                
|    |--- defaultSettings.js             # 项目配置，可设置主题色、异步超时时长、菜单宽度等                
|    |--- global.js                      🔒 全局JS                
|    |--- global.less                    🔒 约定的全局样式，自动引入                 
|--- .editorconfig                       🔒 与支持EditorConfig 插件的IDE协同，达到一个配置适配多种IDE的目的
|--- .eslintignore                       🔒 
|--- .eslintrc.js                        🔒 
|--- .gitignore                          🔒 
|--- .prettierignore                     🔒 
|--- .prettierrc                         🔒
|--- .stylelintrc.json                   🔒
|--- .jsconfig.json                      🔒 
|--- package.json                        🔒 依赖包配置文件
|--- yarn.lock                           🔒 锁定依赖版本号，yarn install时会按里面记录的版本安装 
```

### 命名规范

目录和文件名均采用 __驼峰结构__ 

目录名：pages下存放react组件的，首字母大写

文件名：js类型如果是 __react组件__ 首字母大写(index.js除外)，其他js以及其他类型(less、jpg，ejs)首字母小写。

[更多说明]（https://umijs.org/zh/guide/app-structure.html）

## 路由和菜单

### 路由

`路由管理` 通过约定的语法根据在 _config/router.config.js_ 中配置路由。

`菜单生成` 根据路由配置来生成菜单。菜单项名称，嵌套路径与路由高度耦合。

```javascript
  // 前3项是必填，后两项选填
  {
    path: '/roleManage',
    name: 'roleManage',
    component: '/RoleManage',
    icon: 'usergroup-add',
    hideInMenu: true,
    authority:['admin','guest']
  }
```

`name` 和 `icon`分别代表生成菜单项的文本(name对应国际化文件上的key)和图标。

`authority` 表示当前菜单的权限，如果配置了将会验证当前用户的权限(用户登录成功后 保存在localStorage中的 _antd-pro-authority_ 表示当前用户的权限)，并决定是否展示。

`hideInMenu` true表示当前菜单不会出现在左侧菜单中

### 菜单

菜单根据 _router.config.js_ 生成，具体逻辑在 _src/layouts/BasicLayout_ 中的 `formatter` 方法实现。

如果页面内容是由iframe内容组成的，则需要加上`iframe`属性，后面的值即时url的绝对地址，如下：

配置多个这样的iframe页面，componet 统一指向`IframeWrapper`组件即可

```javascript
      {
        path: '/bd',
        name: 'bd',
        icon: 'dashboard',
        component: '../components/IframeWrapper/index',
        iframe:'http://www.baidu.com'
      },
```

[更多介绍](https://pro.ant.design/docs/router-and-nav-cn)

## 权限管理

[https://pro.ant.design/docs/authority-management-cn](https://pro.ant.design/docs/authority-management-cn)

## 数据mock

`npm run dev` 默认启动的是Mock模式。除了可以模拟POST\GET\请求之前，还可以模拟500、404等错误。

引入[Mock.js](http://mockjs.com/)可以让数据更逼真。

引入[roadhog-api-doc](https://github.com/nikogu/roadhog-api-doc) 可以模拟延迟返回

Mock语法如下:

```javascript
export default {
  // 支持值为 Object 和 Array
  'GET /api/users': { users: [1, 2] },

  // GET POST 可省略
  '/api/users/1': { id: 1 },

  // 支持自定义函数，API 参考 express@4
  'POST /api/users/create': (req, res) => {
    res.status(200).send({
      resultCode: '0',
      resultMsg: '成功',
      resultObject: {
        name: 'Mr.Lin',
        userid: '00000001',
      },
    });
  },
};
```
[更多用法点击这里](https://pro.ant.design/docs/mock-api-cn)

## 国际化

当前暂且只支持`zh-CN`和`en-US`。__建议按模块组织__

配置中文包 _src/locales/zh-CN.js_

```javascript
export default {
  'menu.dashboard': '首页'
}
```

调用

```javascript
import { FormattedMessage} from 'umi/locale';

export default () => {
  return <div><FormattedMessage id="menu.dashboard" /></div>
}
```

可通过`src/defaultSettings.js`中的 _language_ 或 _baseNavigator_ 设置默认语言。

手动切换语言时，会在localStorage中保存 _umi_locale_ 字段。

[更多用法](https://pro.ant.design/docs/i18n-cn)

> 注 ： 在每个异步请求头包含自定义字段language，用于标识当前的语言环境。如：language: 'zh-CN | en-US'

## 数据流方案

[React间组件的通信机制](https://github.com/sunyongjian/blog/issues/27)是单向向下逐级分发，在应对复杂项目时显得捉襟见肘。所以，需要引入一种数据流方案。我们采用的是基于[redux](https://github.com/reduxjs/redux)和[redux-saga](https://github.com/redux-saga/redux-saga)的dva数据流方案。

数据流转描述如下：

1. 用`connect`方法把`model`和`component`链接起来
2. 页面事件触发，开始 `dispatch` `action`
3. `action`到达`model`中的`effects`中间件
4. 中间件获取数据，通过`put`方法把数据流转到指定的`reducer`
5. `reducer`改变state后重新注入组件
6. 组件的props被改变，触发页面重新渲染

![数据流转](https://gw.alipayobjects.com/zos/rmsportal/ZSCxeNAFqHgKXsyjtpxt.png)


model、dispatch、action、reducer和effects都是redux中的重要概念

[Dva概念](https://dvajs.com/guide/concepts.html) | [Dva快速上手](https://www.yuque.com/ant-design/course/abl3ad) | [为什么要用redux](https://segmentfault.com/a/1190000012142449)


## 兼容性

兼容IE9+，以及其他现代浏览器

针对IE9的样式统一定义在 _src/fix-ie9.less_

已知IE9问题

- Upload组件 不支持file.size 无法限制附件上传大小

## 主题配置

可以通过配置`defaultSettings.js`文件快速更改主题效果

```javascript
module.exports = {
  name: 'xxx企业门户', // 项目的名字
  language: 'zh-CN', //表示默认的语言，暂且支持 'zh-CN' 'en-US'。在baseNavigator=true时失效
  baseNavigator: true, //true 表示用navigator.language的值作为默认语言。优先级比language高，比localStorage内的umi_locale低
  timeout: 1000 * 60, //1分钟超时
  footer: 'Copyright©1999-2019 中国移动 版权所有', // footer中显示的字, 可以嵌入html标签
  siderWidth: 200, //侧边栏宽度
  fixSiderbar: true, // sticky siderbar
  delay: 300, //mock接口延迟返回的时长，单位毫秒
  breakpoint: 'xxl', // 根据视窗宽度,自动收起侧边栏的发展，还支持：xs: '480px', sm: '576px',md: '768px',lg: '992px',xl: '1200px',xxl: '1600px'
  navTheme: 'dark', // theme for nav menu
  iconfontUrl:'', //引入Iconfont 自定义图标 用法参考 https://github.com/ant-design/ant-design-pro/pull/3517
}
```

## 异步方法封装

因为需要兼容低版本浏览器IE9，所以放弃使用fetch类的库,（如：Axios,isomorphic-fetch等）。而直接引入jquery，但进行了二次封装。

### 封装内容 

一、约定回参格式，主动提示请求错误。

```javascript
{
  resultCode:"0",
  resultMsg:"成功返归约定数据",
  resultObject:{}
}
```

主动提示会发生于下列两种场景：

1. 异步服务出错时。如：500,404等，提示信息由 _url_, _http code_ 和 _code含义_ 组成。
2. 在resultCode非0时。提示信息来源 _resultMsg_ 。(可通过showError=false 拦截这种提示)

二、扩展了3个属性：_expiry_，_showMask_ 和 _showError_。其中设置 **expiry** 可以缓存数据。根据场景合理应用可以有效 __减少请求，提升用户体验__。

![request](./src/utils/request.gif)

三、对post请求的入参进行了JSON.stringify()处理

四、请求头追加 __language__ 字段，用于标识当前的语言环境。如：language: 'zh-CN | en-US'

### 使用方法

`request(url[,options])` options是一个对象，与jquery.ajax中的[settings](http://www.css88.com/jqapi-1.9/jQuery.ajax/)用法一致。


```javascript
import request from '@/utils/request';

//默认为post请求
request('/api/project/notice')

//get
request('/api/project/notice',{method:'get'})

//带入参
request('/api/project/notice',{data:{name:'Mr.Lin',uid:'8088'}})

//缓存60秒(即60秒内相同的地址和入参请求,不向服务器发请求，直接从sessionStorage取值)
request('/api/project/notice',{expiry:1000*60})

//开启一个全屏遮罩（不推荐这种用法，建议以dva的loading形式实行遮罩）
request('/api/project/notice',{showMask:true})
```

## 动态加载组件

只加载当前页面需要用到的组件，其余的按需加载，对 __性能优化__ 很有意义。

```javascript
import dynamic from 'umi/dynamic';
import LoadingComponent from '@/components/PageLoading/index';

const LoadableRoleInfo = dynamic({
  loader: () => import('./RoleInfo'),
  loading: LoadingComponent,
});
```
`import('./RoleInfo')`  引入组件
`LoadingComponent`  是全局使用的loading效果

通过调用dynamic重新生成一个新的组件，这个组件只会在被调用时，才会异步去取`./RoleInfo`文件

[umi-dynamic](https://umijs.org/zh/api/#umi-dynamic)

## umi约定

- `@` 指向src 目录，如 `@/components`。
- `~`  指向node_modules目录
- 路由表中的component 是相对于 src/pages 目录的。[详解](https://umijs.org/zh/guide/router.html#%E9%85%8D%E7%BD%AE%E5%BC%8F%E8%B7%AF%E7%94%B1)
-  page(s) 文件夹下的 _mock.js 文件即 mock 文件。[详解](https://umijs.org/zh/guide/mock-data.html#%E4%BD%BF%E7%94%A8-umi-%E7%9A%84-mock-%E5%8A%9F%E8%83%BD)
- model 分两类，一是全局 model，二是页面 model。全局 model 存于 /src/models/ 目录，所有页面都可引用；页面 model 不能被其他页面所引用。[详解](https://umijs.org/zh/guide/with-dva.html#model-%E6%B3%A8%E5%86%8C) 

## z-index

关于页面层级的设计

```css
@zindex-table-fixed: auto;
@zindex-affix: 10;
@zindex-back-top: 10;
@zindex-badge: 10;
@zindex-picker-panel: 10;
@zindex-popup-close: 10;
@zindex-modal: 1000;
@zindex-modal-mask: 1000;
@zindex-message: 1010;
@zindex-notification: 1010;
@zindex-popover: 1030;
@zindex-dropdown: 1050;
@zindex-picker: 1050;
@zindex-tooltip: 1060;
```
以上是antd的层级逻辑

以下是关于SiderMenu、Header和Fixed Tabbar的设计

```css
@zindex-sider-menu: 950;
@zindex-header: 900;
@zindex-fixe-tabbar: 900;
```

## 自定义图标

首先，在 __defaultSettings.js__ 配置 `iconfontUrl`，值为iconfont上对应的项目图标库地址，如下：

```javascript
{
  ...
  iconfontUrl:'//at.alicdn.com/t/font_1091403_zssgtcm62kk.js'
}
```

其次，按如下调用。`type`值为对应图标name

```javascript
import IconFont from '@/components/IconFont'

<IconFont type="icon-hongbao" /> 
```
[详细用法](https://github.com/ant-design/ant-design-pro/pull/3517)


## 代码规范

### 规范简介

采用 [_Airbnb_](https://github.com/airbnb/javascript) 代码规范

- [Javascript规范](https://github.com/lin-123/javascript)
- [React代码规范](https://github.com/lin-123/javascript/tree/cn/react)


问题来了，规范写的越详细，开发人员遵守起来就越难。因为这些规范条目就像字典一样，谁也背不下这么多的条目，更不用说在日常开发过程中来运用了。

所以需要引入工具，结合[代码规范](#代码规范)实现 __自动化__ 校验。

### 安装工具

只需要关心IDE插件部分， npm包在yarn install时已经安装并配置完毕 

- vscode 插件
  - `Prettier - Code formatter` 插件
  - `EditorConfig for VS Code` 插件
  - `vscode-stylelint`
- 各种npm包 
  - `babel-eslint` 给eslint指定的解析器，通常react项目都需要这个
  - `prettier` 用来格式化js,less
  - `eslint` 校验模块 需与.eslintrc.js和.eslintignore两个文件配合使用    
    - `eslint-config-prettier` 让eslint继承prettier的配置
    - `eslint-config-airbnb` 让eslint继承airbnb的配置
    - `eslint-plugin-react`
    - `eslint-plugin-jsx-a11y`
    - `eslint-plugin-import`  [airbnb关联的3个模块](https://github.com/airbnb/javascript/tree/4f15da6146fc00096a909e4c53808a3068a4d853/packages/eslint-config-airbnb#eslint-config-airbnb-1)
    - `eslint-plugin-babel` 与babel-eslint配套的ESlint规则插件

changeLog.md
# 2.2.0 (2020-04-28)


### Bug Fixes

* **修复:** 测试修复 ([b3af829](https://gitlab.iwhalecloud.com/soon/react-slick/commit/b3af82988426b709ad2f9cdf666b8e0959c554e6))


### Features

* **blog:** add comment section ([38781f7](https://gitlab.iwhalecloud.com/soon/react-slick/commit/38781f71a5ad99d31f275ee9f0a1ed174f2a2964))



# 2.2.0 (2020-04-28)


### Features

* **blog:** add comment section ([38781f7](https://gitlab.iwhalecloud.com/soon/react-slick/commit/38781f71a5ad99d31f275ee9f0a1ed174f2a2964))



# 2.2.0 (2020-04-28)



# 更新日志

## 2.2.0

`2019-08-12`

- 添加**pubsub-js**模块：订阅发布功能
- **StepWizard**组件添加destroy功能：指定某个场景在被隐藏时，销毁视图，只在**isLazyMount**下有效
- 添加增删改查demo，基于**StepWizard**，**pubsub-js**和**SlickTable**的综合应用

`2019-07-05`

- 升级jquery版本至**jquery@1.12.4**

`2019-06-24`

- 升级 **antd@3.19.1** 至  **antd@3.19.6**
- 添加linter插件 **eslint-plugin-react-hooks** 
- 添加 **@/components/SlickUpload** 
- **global.less** 添加utils样式 

`2019-06-17`

- 优化 **@/components/SlickTable** 滚动条仅在过长时显示；修复pick模式下 水平滚动时左边列显示不完整
- 优化 **@/components/IframeWrapper** 
- 优化 **SceneSwitch2** demo 
- 屏蔽冗余的demo
- 增加 **getViewPort** 工具方法

`2019-06-14`

- 修复 **@/components/FooterToolbar** IE下 card切换时 依然可见的bug

## 2.1.3

`2019-06-10`

- 添加组件 **@/components/SlickTable**

`2019-06-04`

- 添加组件 **@/components/StepWizard**
- 添加 **StepWizard** demo
- eslint: 停止校验析构赋值 **react/destructuring-assignment**

`2019-05-31`

- 调整eslint规则：1、允许变量声明与外层作用域的变量同名；2、剔除所有mock 文件的校验

`2019-05-30`

- 移除basicLayout 冗余的console.log信息
- request优化：请求头添加language字段，并在readme补充国际化说明

## 2.1.2

`2019-05-29`

- readme文案优化
- 移除冗余的types目录
- 修复IEVersion方法
- 升级到antd@3.19.1，解决
- 修复IE9下Tabs无法正常切换(animated属性引起)

## 2.1.1

`2019-05-21`

- prettier遗留的没被格式化过的js、less文件 

## 2.1.0

`2019-05-20`

- 修改**components/ScrollBar** 滚动条颜色
- 优化：tab切换和关闭时即时更新url，同时修复路由权限问题
- 添加eslint自动化校验功能

## 2.0.1

`2019-04-19`

- 新增自定义图标组件IconFont
- 新增**react-custom-scrollbars**模块
- 优化SiderMenu滚动条

## 2.0.0

`2019-04-17`

- 新增分支drive-by-zeal-cube ，该分支引入的是zeal-cube组件
- 把源码中对'zeal-cube'的引用切换回'antd',移除config.js和basicLayout.js中的国际化配置。同时，移除package.json中的zeal-cube
- 升级umi至2.6.13, 
- 升级umi-plugin-react至1.7.4。
- react更新到16.8.3
- antd更新到3.16.3
- 调整siderMenu和header布局
- 拆分路由和国际化
- 更新pages/demo/sceneSwith和pages/Dashboard(旧首页更名为Dashboard2)

## 1.0.0

## 添加TableWrapper组件

统一table风格

demo : http://localhost:8000/#/log/detail

## 添加ComboCascader组件

demo : http://localhost:8000/#/form/inputs

## 添加SearchFilter组件

demo : http://localhost:8000/#/form/inputs

## 添加.cube-form-tooltip

扩展表单校验的错误样式

demo : http://localhost:8000/#/form/inputs

## 扩展card组件 样式：添加.cube-tabs-filled 和 .cube-tabs-lined

demo : http://localhost:8000/#/roleManage

## 扩展btn样式cube-btn-info和cube-btn-round

demo : http://localhost:8000/#/buttons






mock/api.js
import mockjs from 'mockjs';
import { delay } from 'roadhog-api-doc';
import defaultSettings from '../src/defaultSettings';

/**
 * 从 www.abc.com?id=1&name=Mr.Lin 字符串中提取 对应key的值
 * @ paramName {String}
 * @ searchString {String}
 */
function getURLParameter(paramName, searchString) {
  const str = searchString.split('?')[1];
    let i;
    let val;
    const params = str.split('&');

  for (i = 0; i < params.length; i++) {
    val = params[i].split('=');
    if (val[0] === paramName) {
      return decodeURIComponent(val[1]);
    }
  }
  return null;
}
function getUniqueID(prefix) {
  return `${prefix  }_${  Math.floor(Math.random() * (new Date()).getTime())}`;
}
const titles = [
  'Alipay',
  'Angular',
  'Ant Design',
  'Ant Design Pro',
  'Bootstrap',
  'React',
  'Vue',
  'Webpack',
];
const avatars = [
  'https://gw.alipayobjects.com/zos/rmsportal/WdGqmHpayyMjiEhcKoVE.png', // Alipay
  'https://gw.alipayobjects.com/zos/rmsportal/zOsKZmFRdUtvpqCImOVY.png', // Angular
  'https://gw.alipayobjects.com/zos/rmsportal/dURIMkkrRFpPgTuzkwnB.png', // Ant Design
  'https://gw.alipayobjects.com/zos/rmsportal/sfjbOqnsXXJgNCjCzDBL.png', // Ant Design Pro
  'https://gw.alipayobjects.com/zos/rmsportal/siCrBXXhmvTQGWPNLBow.png', // Bootstrap
  'https://gw.alipayobjects.com/zos/rmsportal/kZzEzemZyKLKFsojXItE.png', // React
  'https://gw.alipayobjects.com/zos/rmsportal/ComBAopevLwENQdKWiIn.png', // Vue
  'https://gw.alipayobjects.com/zos/rmsportal/nxkuOJlFJuAUhzlMTCEe.png', // Webpack
];

const avatars2 = [
  'https://gw.alipayobjects.com/zos/rmsportal/BiazfanxmamNRoxxVxka.png',
  'https://gw.alipayobjects.com/zos/rmsportal/cnrhVkzwxjPwAaCfPbdc.png',
  'https://gw.alipayobjects.com/zos/rmsportal/gaOngJwsRYRaVAuXXcmB.png',
  'https://gw.alipayobjects.com/zos/rmsportal/ubnKSIfAJTxIgXOKlciN.png',
  'https://gw.alipayobjects.com/zos/rmsportal/WhxKECPNujWoWEFNdnJE.png',
  'https://gw.alipayobjects.com/zos/rmsportal/jZUIxmJycoymBprLOUbT.png',
  'https://gw.alipayobjects.com/zos/rmsportal/psOgztMplJMGpVEqfcgF.png',
  'https://gw.alipayobjects.com/zos/rmsportal/ZpBqSxLxVEXfcUNoPKrz.png',
  'https://gw.alipayobjects.com/zos/rmsportal/laiEnJdGHVOhJrUShBaJ.png',
  'https://gw.alipayobjects.com/zos/rmsportal/UrQsqscbKEpNuJcvBZBu.png',
];

const covers = [
  'https://gw.alipayobjects.com/zos/rmsportal/uMfMFlvUuceEyPpotzlq.png',
  'https://gw.alipayobjects.com/zos/rmsportal/iZBVOIhGJiAnhplqjvZW.png',
  'https://gw.alipayobjects.com/zos/rmsportal/iXjVmWVHbCJAyqvDxdtx.png',
  'https://gw.alipayobjects.com/zos/rmsportal/gLaIAoVWTtLbBWZNYEMg.png',
];
const desc = [
  '那是一种内在的东西， 他们到达不了，也无法触及的那是一种内在的东西， 他们到达不了，也无法触及的那是一种内在的东西， 他们到达不了，也无法触及的',
  '希望是一个好东西，也许是最好的，好东西是不会消亡的',
  '生命就像一盒巧克力，结果往往出人意料',
  '城镇中有那么多的酒馆，她却偏偏走进了我的酒馆',
  '那时候我只会想自己想要什么，从不想自己拥有什么',
];

const user = [
  '付小小',
  '曲丽丽',
  '林东东',
  '周星星',
  '吴加好',
  '朱偏右',
  '鱼酱',
  '乐哥',
  '谭小仪',
  '仲尼',
];

function fakeList(count) {
  const list = [];
  for (let i = 0; i < count; i += 1) {
    list.push({
      id: `fake-list-${i}`,
      owner: user[i % 10],
      title: titles[i % 8],
      avatar: avatars[i % 8],
      cover: parseInt(i / 4, 10) % 2 === 0 ? covers[i % 4] : covers[3 - (i % 4)],
      status: ['active', 'exception', 'normal'][i % 3],
      percent: Math.ceil(Math.random() * 50) + 50,
      logo: avatars[i % 8],
      href: 'https://ant.design',
      updatedAt: new Date(new Date().getTime() - 1000 * 60 * 60 * 2 * i),
      createdAt: new Date(new Date().getTime() - 1000 * 60 * 60 * 2 * i),
      subDescription: desc[i % 5],
      description:
        '在中台产品的研发过程中，会出现不同的设计规范和实现方式，但其中往往存在很多类似的页面和组件，这些类似的组件会被抽离成一套标准规范。',
      activeUser: Math.ceil(Math.random() * 100000) + 100000,
      newUser: Math.ceil(Math.random() * 1000) + 1000,
      star: Math.ceil(Math.random() * 100) + 100,
      like: Math.ceil(Math.random() * 100) + 100,
      message: Math.ceil(Math.random() * 10) + 10,
      content:
        '段落示意：蚂蚁金服设计平台 ant.design，用最小的工作量，无缝接入蚂蚁金服生态，提供跨越设计与开发的体验解决方案。蚂蚁金服设计平台 ant.design，用最小的工作量，无缝接入蚂蚁金服生态，提供跨越设计与开发的体验解决方案。',
      members: [
        {
          avatar: 'https://gw.alipayobjects.com/zos/rmsportal/ZiESqWwCXBRQoaPONSJe.png',
          name: '曲丽丽',
          id: 'member1',
        },
        {
          avatar: 'https://gw.alipayobjects.com/zos/rmsportal/tBOxZPlITHqwlGjsJWaF.png',
          name: '王昭君',
          id: 'member2',
        },
        {
          avatar: 'https://gw.alipayobjects.com/zos/rmsportal/sBxjgqiuHMGRkIjqlQCd.png',
          name: '董娜娜',
          id: 'member3',
        },
      ],
    });
  }

  return list;
}

let sourceData;

function postFakeList(req, res) {
  const { /* url = '', */ body } = req;
  // const params = getUrlParams(url);
  const { method, id } = body;
  // const count = (params.count * 1) || 20;
  let result = sourceData;

  switch (method) {
    case 'delete':
      result = result.filter(item => item.id !== id);
      break;
    case 'update':
      result.forEach((item, i) => {
        if (item.id === id) {
          result[i] = Object.assign(item, body);
        }
      });
      break;
    case 'post':
      result.unshift({
        body,
        id: `fake-list-${result.length}`,
        createdAt: new Date().getTime(),
      });
      break;
    default:
      break;
  }

  return res.json(result);
}

const getNotice = [
  {
    id: 'xxx1',
    title: titles[0],
    logo: avatars[0],
    description: '那是一种内在的东西，他们到达不了，也无法触及的',
    updatedAt: new Date(),
    member: '科学搬砖组',
    href: '',
    memberLink: '',
  },
  {
    id: 'xxx2',
    title: titles[1],
    logo: avatars[1],
    description: '希望是一个好东西，也许是最好的，好东西是不会消亡的',
    updatedAt: new Date('2017-07-24'),
    member: '全组都是吴彦祖',
    href: '',
    memberLink: '',
  },
  {
    id: 'xxx3',
    title: titles[2],
    logo: avatars[2],
    description: '城镇中有那么多的酒馆，她却偏偏走进了我的酒馆',
    updatedAt: new Date(),
    member: '中二少女团',
    href: '',
    memberLink: '',
  },
  {
    id: 'xxx4',
    title: titles[3],
    logo: avatars[3],
    description: '那时候我只会想自己想要什么，从不想自己拥有什么',
    updatedAt: new Date('2017-07-23'),
    member: '程序员日常',
    href: '',
    memberLink: '',
  },
  {
    id: 'xxx5',
    title: titles[4],
    logo: avatars[4],
    description: '凛冬将至',
    updatedAt: new Date('2017-07-23'),
    member: '高逼格设计天团',
    href: '',
    memberLink: '',
  },
  {
    id: 'xxx6',
    title: titles[5],
    logo: avatars[5],
    description: '生命就像一盒巧克力，结果往往出人意料',
    updatedAt: new Date('2017-07-23'),
    member: '骗你来学计算机',
    href: '',
    memberLink: '',
  },
];

const getActivities = [
  {
    id: 'trend-1',
    updatedAt: new Date(),
    user: {
      name: '曲丽丽',
      avatar: avatars2[0],
    },
    group: {
      name: '高逼格设计天团',
      link: 'http://github.com/',
    },
    project: {
      name: '六月迭代',
      link: 'http://github.com/',
    },
    template: '在 @{group} 新建项目 @{project}',
  },
  {
    id: 'trend-2',
    updatedAt: new Date(),
    user: {
      name: '付小小',
      avatar: avatars2[1],
    },
    group: {
      name: '高逼格设计天团',
      link: 'http://github.com/',
    },
    project: {
      name: '六月迭代',
      link: 'http://github.com/',
    },
    template: '在 @{group} 新建项目 @{project}',
  },
  {
    id: 'trend-3',
    updatedAt: new Date(),
    user: {
      name: '林东东',
      avatar: avatars2[2],
    },
    group: {
      name: '中二少女团',
      link: 'http://github.com/',
    },
    project: {
      name: '六月迭代',
      link: 'http://github.com/',
    },
    template: '在 @{group} 新建项目 @{project}',
  },
  {
    id: 'trend-4',
    updatedAt: new Date(),
    user: {
      name: '周星星',
      avatar: avatars2[4],
    },
    project: {
      name: '5 月日常迭代',
      link: 'http://github.com/',
    },
    template: '将 @{project} 更新至已发布状态',
  },
  {
    id: 'trend-5',
    updatedAt: new Date(),
    user: {
      name: '朱偏右',
      avatar: avatars2[3],
    },
    project: {
      name: '工程效能',
      link: 'http://github.com/',
    },
    comment: {
      name: '留言',
      link: 'http://github.com/',
    },
    template: '在 @{project} 发布了 @{comment}',
  },
  {
    id: 'trend-6',
    updatedAt: new Date(),
    user: {
      name: '乐哥',
      avatar: avatars2[5],
    },
    group: {
      name: '程序员日常',
      link: 'http://github.com/',
    },
    project: {
      name: '品牌迭代',
      link: 'http://github.com/',
    },
    template: '在 @{group} 新建项目 @{project}',
  },
];

function getFakeCaptcha(req, res) {
  return res.json('captcha-xxx');
}

const proxy = {
  'GET /api/project/notice': getNotice,
  'GET /api/activities': getActivities,
  'POST /api/forms': (req, res) => {
    res.send({
      resultCode: '0',
      resultMsg: '操作成功!',
      resultObject: {},
    });
  },
  'GET /api/tags': mockjs.mock({
    'list|100': [{ name: '@city', 'value|1-100': 150, 'type|0-2': 1 }],
  }),
  'GET /api/fake_list': (req, res) => {    
    const count = getURLParameter('count', req.url) * 1 || 20;
    const result = fakeList(count);
    sourceData = result;
    res.status(200).send({
      resultCode: '0',
      resultMsg: '操作成功!',
      resultObject: result,
    });
  },
  'POST /api/fake_list': postFakeList,
  'GET /api/captcha': getFakeCaptcha,
  'POST /api/area': (req, res) => {
    res.send(
      mockjs.mock({
        resultCode: '0',
        resultMsg: '操作成功!',
        resultObject: [
          {
            label: '@city()',
            value: getUniqueID('v'),
            fieldname: getUniqueID('p'),
            isLeaf: true
          },
          {
            label: '@city()',
            value: getUniqueID('v'),
            fieldname: getUniqueID('p')
          },
        ],
      })
      );
  },
};

export default delay(proxy, defaultSettings.delay);